stablehlo/reference/Ops.cpp

```diff
@@ -618,6 +770,58 @@ SmallVector<Tensor> eval(
       Tensor runtimeOperand = scope.find(realOp.getOperand());
       Tensor runtimeResult = evalRealOp(runtimeOperand, realOp.getType());
       scope.add(op.getResults(), {runtimeResult});
+    } else if (auto reduceOp = dyn_cast<ReduceOp>(op)) {
+      SmallVector<Tensor> runtimeInputs = scope.find(reduceOp.getInputs());
+      SmallVector<Tensor> runtimeInitValues =
+          scope.find(reduceOp.getInitValues());
+      SmallVector<TensorType> resultTypes;
+      for (auto resultType : reduceOp.getResultTypes())
+        resultTypes.push_back(resultType.cast<TensorType>());
+      SmallVector<Tensor> runtimeResults = evalReduceOp(
+          runtimeInputs, runtimeInitValues, Axes(reduceOp.getDimensions()),
+          reduceOp.getBody(), scope, resultTypes);
+      scope.add(op.getResults(), {runtimeResults});
+    } else if (auto reduceWindowOp = dyn_cast<ReduceWindowOp>(op)) {
+      SmallVector<Tensor> runtimeInputs =
+          scope.find(reduceWindowOp.getInputs());
+      SmallVector<Tensor> runtimeInitValues =
+          scope.find(reduceWindowOp.getInitValues());
+      SmallVector<TensorType> resultTypes;
+      for (auto resultType : reduceWindowOp.getResultTypes())
+        resultTypes.push_back(resultType.cast<TensorType>());
+      SmallVector<int64_t> windowStrides =
+          SmallVector<int64_t>(runtimeInputs[0].getRank(), 1);
+      SmallVector<int64_t> baseDilations =
+          SmallVector<int64_t>(runtimeInputs[0].getRank(), 1);
+      SmallVector<int64_t> windowDilations =
+          SmallVector<int64_t>(runtimeInputs[0].getRank(), 1);
+      SmallVector<SmallVector<int64_t>> padding =
+          SmallVector<SmallVector<int64_t>>(runtimeInputs[0].getRank(), {0, 0});
+      if (reduceWindowOp.getWindowStrides().has_value())
+        windowStrides = llvm::to_vector(
+            (*reduceWindowOp.getWindowStrides()).getValues<int64_t>());
+      if (reduceWindowOp.getBaseDilations().has_value())
+        baseDilations = llvm::to_vector(
+            (*reduceWindowOp.getBaseDilations()).getValues<int64_t>());
+      if (reduceWindowOp.getWindowDilations().has_value())
+        windowDilations = llvm::to_vector(
+            (*reduceWindowOp.getWindowDilations()).getValues<int64_t>());
+      if (reduceWindowOp.getPadding().has_value()) {
```

Comment:
padding can use the same type as it is returned from `hlo::convertPaddingAttribute` so that we do not need any conversion code:

```
  SmallVector<std::pair<int64_t, int64_t>> padding(
          runtimeInputs[0].getRank(), std::pair<int64_t, int64_t>(0, 0));
      if (reduceWindowOp.getPadding().has_value()) {
        auto paddingOrErr =
            hlo::convertPaddingAttribute(reduceWindowOp.getPadding(), {});
        if (failed(paddingOrErr))
          report_fatal_error(invalidArgument("Invalid padding format found."));
        padding = *paddingOrErr;
      }

```